### PR TITLE
Build with only montgomery curves (+ DJB configuration)

### LINKED ISSUE #941: Configuration issue with using only MONTGOMERY and no SHORT_WEIERSTRASS
[mbedtls_config.txt](https://github.com/ARMmbed/mbedtls/files/1035992/mbedtls_config.txt)

I was creating my own config.h. You can find it attached. When I build my application I get

    In function `mbedtls_ecp_muladd': 
    ./libs/mbedtls-2.4.2/library/ecp.c:1739: undefined reference to `ecp_add_mixed'
    ./libs/mbedtls-2.4.2/library/ecp.c:1740: undefined reference to `ecp_normalize_jac'

Both `ecp_add_mixed` and `ecp_normalize_jac` are only defined if `ECP_SHORTWEIERSTRASS` is defined. `mbedtls_ecp_muladd` itself returns with an error code if the group type is not equal to `ECP_TYPE_SHORT_WEIERSTRASS`. 

I therefore assume that the whole function can be put into the 
 
    #if defined(ECP_SHORTWEIERSTRASS)
    ...
    #endif

block beginning in line 1431 of ecp.c (and probably also `mbedtls_ecp_mul_shortcuts`. Am I correct? Or is there a reason why this is like it is?

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Fix compile-time checks to allow building with ECP enabled, but no short Weierstrass curve (i.e. with only Montgomery curves: Curve25519 or Curve448). Fix #941, #1412.

I wasn't sure how to frame a non-regression test. `scripts/curves.pl` didn't catch this because it tries building with all-curves-but-one. This is the right way to do what this script was designed to do, which is to catch test cases that are missing a dependency on a curve, including tests that depend on multiple curves. Maybe we should also try building with a single curve, for each curve?

What I've done in this PR is to add a new reference configuration with only “DJB” crypto and NIST hashes: SHA-256, SHA-512, HMAC_DRBG, Chacha/Poly, and ECDH/ECDSA with Curve25519. I'm not sure if that configuration makes sense for the time being though, since we don't fully support Curve25519 for ECDSA and TLS. Suggestions welcome.

### CODE REVIEW

File: tests/suites/test_suite_cipher.chacha20.data [hunk 1]
```diff
@@ -1,7 +1,3 @@
-Decrypt empty buffer
-depends_on:MBEDTLS_CHACHA20_C
-dec_empty_buf:
```
Comment (Patater) [inline tests/suites/test_suite_cipher.chacha20.data]: I don't understand why we can't have a `dec_empty_buf` test anymore. Should we instead parameterize the test to so it uses a cipher parameter?

It seems like we are reducing the intended test coverage here by removing this test. (I think we intended to have an empty buffer test that used CHACHA20.)

File: programs/ssl/ssl_client2.c [hunk 2]
```diff
@@ -39,12 +39,22 @@
 
 #if !defined(MBEDTLS_ENTROPY_C) || \
     !defined(MBEDTLS_SSL_TLS_C) || !defined(MBEDTLS_SSL_CLI_C) || \
-    !defined(MBEDTLS_NET_C) || !defined(MBEDTLS_CTR_DRBG_C)
+    !defined(MBEDTLS_NET_C)
```
Comment (Patater) [inline programs/ssl/ssl_client2.c]: What does `Support HMAC_DRBG in the TLS test programs ` have to do with `Build with only montgomery curves`? Shouldn't this commit be in a different PR?

**Comment by Patater** [comment]
I like the approach to testing this by adding a new configuration with "only 'DJB' crypto and NIST hashes".

File: library/ecp.c [hunk 3]
```diff
@@ -1845,6 +1848,16 @@ int mbedtls_ecp_muladd( mbedtls_ecp_group *grp, mbedtls_ecp_point *R,
...
     return( ret );
+
+#else /* ECP_SHORTWEIERSTRASS */
+    (void) grp;
+    (void) R;
+    (void) m;
+    (void) P;
+    (void) n;
+    (void) Q;
+    return( MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE );
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/ecp.c]: Because the function existed and I don't want to remove a function in a minor version.

File: tests/suites/test_suite_cipher.chacha20.data (hunk 1 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_cipher.chacha20.data]: BTW I don't think DJB likes HMAC_DRBG either, but CTR_DRBG and HMAC_DRBG are all we have. HMAC_DRBG builds on SHA2 which is pretty much indispensable anyway.

Comment (mpg) [reply tests/suites/test_suite_cipher.chacha20.data]: I agree with Gilles that removing the test does not reduce the actual test coverage (which might indeed differ from the intended one, though I'm not sure our intentions here were very clear), so I think it would better be handled as a separate issue, in order to avoid feature creep in this one.

> BTW I don't think DJB likes HMAC_DRBG either, but CTR_DRBG and HMAC_DRBG are all we have. HMAC_DRBG builds on SHA2 which is pretty much indispensable anyway.

I assume DJB would prefer a DRBG based on ChaCha20, but as you say we don't have it (and I'm not sure one is standardized, so that could be risky to implement), and I think SHA-2 is the part of NIST Crypto that is also approved by DJB (used in Ed25519 for example), so as you say, since HMAC_DRBG builds on it... Btw, from my perspective the main benefit of a DRBG based on Chacha20 rather than HMAC would be speed, as otherwise both should have similar security properties.

File: library/ecp.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (mpg) [reply library/ecp.c]: I'm don't think compatibility considerations apply here, as they would only matter for a non-default configuration that was simply not possible before (and never was). Prior to that change, with only Mongomery curves, you couldn't build ecp.c, so you didn't have that function either. After that change, with only Montgomery curves, you would start having the rest of the module but no that function, which you didn't have anyway. So I think I would prefer compile-time exclusion here.

Note that removing this function at compile-time also means making ECDSA dependent on at least one ECDSA-capable curve (currently, only Short Weierstrass curves) to be defined, because ECDSA uses muladd().

File: configs/config-montgomery.h [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+/* Save RAM at the expense of ROM */
+#define MBEDTLS_AES_ROM_TABLES
+
+/* Save RAM by adjusting to our exact needs */
+#define MBEDTLS_ECP_MAX_BITS   256
+#define MBEDTLS_MPI_MAX_SIZE    32 // 256 bits is 32 bytes
+
+/* Save RAM at the expense of speed, see ecp.h */
+#define MBEDTLS_ECP_WINDOW_SIZE        2
+#define MBEDTLS_ECP_FIXED_POINT_OPTIM  0
```
Comment (mpg) [inline configs/config-montgomery.h]: Only applicable to short-Weierstrass curves. Again, not technically relevant, but could be confusing for humans.

File: configs/config-montgomery.h [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+/*
+ * You should adjust this to the exact number of sources you're using: default
+ * is the "mbedtls_platform_entropy_poll" source, but you may want to add other ones.
+ * Minimum is 2 for the entropy test suite.
+ */
+#define MBEDTLS_ENTROPY_MAX_SOURCES 2
+
+/* Save ROM and a few bytes of RAM by specifying our own ciphersuite list */
+#define MBEDTLS_SSL_CIPHERSUITES                        \
+    MBEDTLS_TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256, \
```
Comment (mpg) [inline configs/config-montgomery.h]: Note: this one will need more complete support of RFC 8422 (EdDSA), and also support of https://tools.ietf.org/html/draft-ietf-curdle-pkix-10 for the certificate chain if I'm not mistaken (unless of course one only uses explicitly-trusted self-signed certificates, in which case one would be better off using raw public keys, but we don't support that yet).

File: tests/suites/test_suite_cipher.chacha20.data (hunk 1 above)
Comment (Patater) [reply tests/suites/test_suite_cipher.chacha20.data]: OK, noted as separate issue in #2014

**Comment by gilles-peskine-arm** [comment]
For several reasons, I've changed my mind as to how to test this.

* Defining an ad hoc configuration lets us test one specific configuration that is of some interest. However, the configuration in question is of limited interest, because you can't use Montgomery curves for anything directly useful yet: the ECDH and ECDSA module don't support them, and they can't be used in TLS at all.
* We normally keep supporting the sample configurations for a long time. The configuration in `config-montgomery.h` is fairly ad hoc and is likely to evolve as our support for Montgomery curves improves.
* The limited configuration in `config-montgomery.h` didn't catch that the ECP self-test was failing since it didn't include `MBEDTLS_SELF_TEST`. This may happen with other options as well.

I think I'll switch to my other idea: modify `curves.pl` to test each single-curve configuration in addition to testing each all-curves-but-one configuration.

**Comment by gilles-peskine-arm** [comment]
Superseded by #2019.

#2019 doesn't have "Support HMAC_DRBG in the TLS test programs". I'll make a separate PR for that.
